_posts/2020-03-10-serverless-express-sls.md

```diff
@@ -0,0 +1,485 @@
+
+<center><img src="/img/express-sls/xray-latency.png" /></center>
+<br/>
+
+We can list all those traces...
+
+<center><img src="/img/express-sls/xray-traces.png" /></center>
+<br/>
+
+And review a trace for a specific request.
+
+<center><img src="/img/express-sls/xray-trace-get.png" /></center>
+<br/>
+
+For those who followed our previous installment, you'll remember that to get all of the comments originally took around 1.6 seconds, when each comment was stored in S3. This is significantly faster at 60ms per request. That's 26 times faster, which is a pretty big improvement. The moral of the story is to perhaps not use S3 in such a scenario - use DynamoDB.
+
+# Other Considerations and Commentary
+
+HTTP API's do not yet have all the features of REST API's. This largely seems to be due to the result of customer feedback; a simple lambda proxy that provides JWT authorization covers a fairly large number of use-cases. Still, it's worth some additional discussion.
+
+> HTTP API vs REST API
+
+At the moment HTTP API's do not support direct service integrations but that are probably on the roadmap. AWS has stated that HTTP API's will eventually hit feature parity with REST API's. Note as well, the performance improvement was derived mainly from switching out S3 for DynamoDB - in a future installment, I will do a more Apples-to-Apples comparision of REST API vs HTTP API. AWS has claimed that HTTP API's are 60% faster than REST APIs, so I expect that HTTP API with Lambda will have comparable performance to REST APIs with Service Integrations, at least for this application.
+
+> Authentication - IAM vs JWT
+
+The serverless express applications used JWTs authorization because it is all that HTTP API's support. REST API's have a more robust selection. In this case I chose to use IAM Authorization. I personally prefer native IAM controls, because it lets me piggy-back on to a more robust RBAC mechanism that I do not need to write myself. In practice this can make things complicated, because in practice it can require using cognito identity pools to vend out AWS credentials via a token exchange. As mentioned in prior installments, Casbin can be used if you want to stick to just using JWTs.
+
+> OpenAPI Document Pollution
+
+It often bothers people that they must include amazon extensions in the OpenAPI document. I can understand wanting to keep the document 'pure' from vendor pollution. To do this, it is possible to define the vendor extensions in a separate file, and then merge the two files afterwards as part of your build process. Alternatively, AWS have their own IDL, called [Smithy](). Smithy can be used to generate an OpenAPI definition file with and without API Gateway extensions. Some users may find this useful if they want to publish their OpenAPI document free from vendor properties that may expose implementation details. I plan to go into specifics in a future installment.
```

Comment:
I plan to go into specifics in a future installment.

I thought this was the end?

Reply:
There's two around REST APIs - this one and the IAM authorizer one.

Then there's one with is more 'best practice' use of HTTP API, which will include a shallow comparison with the other methods

I'm debating whether to do one on the Smithy IDL - because it's a bit of pain to use.

Then I might do a generic 'retrospective' across all the articles and make suggestions on when to choose each type of API.

So that's potentially 3 or 4 more on this track.

I was considering doing something CI/CD related, e.g. blue green traffic shifting - but I might do that as part of the forthcoming CI/CD series instead.